Source/WebKit/UIProcess/Extensions/API/WebExtensionContextAPIStorage.cpp

```diff
@@ -184,38 +197,51 @@
     completionHandler({ });
 }
 
-void WebExtensionContext::fireStorageChangedEventIfNeeded(NSDictionary *oldKeysAndValues, NSDictionary *newKeysAndValues, WebExtensionDataType dataType)
+void WebExtensionContext::fireStorageChangedEventIfNeeded(HashMap<String, String> oldKeysAndValues, HashMap<String, String> newKeysAndValues, WebExtensionDataType dataType)
 {
-    static NSString * const newValueKey = @"newValue";
-    static NSString * const oldValueKey = @"oldValue";
+    static constexpr auto newValueKey = "newValue"_s;
+    static constexpr auto oldValueKey = "oldValue"_s;
 
-    if (!oldKeysAndValues.count && !newKeysAndValues.count)
+    if (!oldKeysAndValues.size() && !newKeysAndValues.size())
         return;
 
-    auto *changedData = [NSMutableDictionary dictionary];
+    RefPtr changedData = JSON::Object::create();
+
+    // Process new or changed keys
+    for (auto mapData : newKeysAndValues) {
+        auto key = mapData.key;
+        auto value = mapData.value;
 
-    // Process new or changed keys.
-    [newKeysAndValues enumerateKeysAndObjectsUsingBlock:^(NSString *key, NSString *newValue, BOOL *) {
-        NSString *oldValue = oldKeysAndValues[key];
+        String oldValue = oldKeysAndValues.get(key);
 
-        if (!oldValue || ![oldValue isEqualToString:newValue]) {
-            id parsedNewValue = parseJSON(newValue, JSONOptions::FragmentsAllowed) ?: NSNull.null;
-            id parsedOldValue = oldValue ? parseJSON(oldValue, JSONOptions::FragmentsAllowed) ?: NSNull.null : nil;
-            changedData[key] = parsedOldValue ? @{ oldValueKey: parsedOldValue, newValueKey: parsedNewValue } : @{ newValueKey: parsedNewValue };
+        if (oldValue.isEmpty() || oldValue != value) {
+            RefPtr parsedNewValue = JSON::Value::parseJSON(value);
+            RefPtr parsedOldValue = JSON::Value::parseJSON(oldValue);
+            RefPtr data = JSON::Object::create();
+            if (parsedOldValue)
+                data->setValue(oldValueKey, *parsedOldValue);
+            data->setValue(newValueKey, *parsedNewValue);
+            changedData->setObject(key, *data);
         }
-    }];
+    }
 
     // Process removed keys.
-    [oldKeysAndValues enumerateKeysAndObjectsUsingBlock:^(NSString *key, NSString *oldValue, BOOL *) {
-        if (!newKeysAndValues[key])
-            changedData[key] = @{ oldValueKey: parseJSON(oldValue, JSONOptions::FragmentsAllowed) ?: NSNull.null };
-    }];
+    for (auto mapData : oldKeysAndValues) {
+        auto key = mapData.key;
+        auto value = mapData.value;
```

Comment:
I like `entry` for these iterator vars.
```suggestion
    for (auto entry : oldKeysAndValues) {
        const auto& key = entry.key;
        const auto& value = entry.value;
```